api/uniter/uniter.go

```diff
@@ -395,6 +395,51 @@ func (st *State) AllMachinePorts(machineTag names.MachineTag) (map[network.PortR
 	return portsMap, nil
 }
 
+// OpenedMachinePortRanges returns all port ranges currently open on the given
+// machine, grouped by unit tag and application endpoint.
+// relation that applies.
+func (st *State) OpenedMachinePortRanges(machineTag names.MachineTag) (map[names.UnitTag]map[string][]network.PortRange, error) {
+	if st.BestAPIVersion() < 17 {
+		// OpenedMachinePortRanges() was introduced in UniterAPIV17.
+		return nil, errors.NotImplementedf("AllMachinePorts() (need V17+)")
+	}
+	var results params.OpenMachinePortRangesResults
+	args := params.Entities{
+		Entities: []params.Entity{{Tag: machineTag.String()}},
+	}
+	err := st.facade.FacadeCall("OpenedMachinePortRanges", args, &results)
+	if err != nil {
+		return nil, err
+	}
+	if len(results.Results) != 1 {
+		return nil, fmt.Errorf("expected 1 result, got %d", len(results.Results))
+	}
+	result := results.Results[0]
+	if result.Error != nil {
+		return nil, result.Error
+	} else if result.GroupKey != "endpoint" {
+		return nil, fmt.Errorf("expected open unit port ranges to be grouped by endpoint, got %s", result.GroupKey)
+	}
+
+	portRangeMap := make(map[names.UnitTag]map[string][]network.PortRange)
```

Comment:
A map of a map of a slice seems a bit complicated to deal with.  Is there a way to easier to use like has been done with the network addresses etc?  Or types?

Reply:
I couldn't find a better way to organize this information. I _could_ introduce a helper type here but then we would get references to api/uniter.$thing in the uniter's runner/context package which looks even worse (the uniter stores this information in the context)... So instead, I opted to use core types.

However, my next PR does introduce a helper type (essentially replaces ports.go in the runner/context) which gets initialized with this map and then offers an API for working with it in a sane manner.